prefect_kubernetes/utilities.py

```diff
@@ -180,3 +202,123 @@ def _slugify_label_value(value: str, max_length: int = 63) -> str:
     # Kubernetes to throw the validation error
 
     return slug
+
+
+class ResilientStreamWatcher:
+    """
+    A wrapper class around kuberenetes.watch.Watch that will reconnect on
+    certain exceptions.
+    """
+
+    DEFAULT_RECONNECT_EXCEPTIONS = (urllib3.exceptions.ProtocolError,)
+
+    def __init__(
+        self,
+        logger: Optional[logging.Logger] = None,
+        max_cache_size: int = 50000,
+        reconnect_exceptions: Optional[List[Type[Exception]]] = None,
+    ) -> None:
+        """
+        A utility class for managing streams of Kuberenetes API objects and logs
+
+        Attributes:
+            logger: A logger which will be used interally to log errors
+            max_cache_size: The maximum number of API objects to track in an
+                internal cache to help deduplicate results on stream reconnects
+            reconnect_exceptions: A list of exceptions that will cause the stream
+                to reconnect.
+        """
+
+        self.max_cache_size = max_cache_size
+        self.logger = logger
+        self.watch = watch.Watch()
+
+        reconnect_exceptions = (
+            reconnect_exceptions
+            if reconnect_exceptions is not None
+            else self.DEFAULT_RECONNECT_EXCEPTIONS
+        )
+        self.reconnect_exceptions = tuple(reconnect_exceptions)
+
+    def stream(self, func: Callable, *args, cache: Optional[Set] = None, **kwargs):
+        """
+        A private method for streaming API objects or logs from a Kubernetes
+        client function. This method will reconnect the stream on certain
+        configurable exceptions and deduplicate results on reconnects if
+        streaming API objects and a cache is provided.
+
+        Note that client functions that produce a stream of logs will
+        restart a stream from the beginning of the log's history on reconnect.
+        It is possible for duplicate logs to be yielded.
```

Comment:
I think we can make this section a little more generic now.
```suggestion
        Note that client functions that produce a stream will
        restart a stream from the beginning of the log's history on reconnect.
        If a cache is not provided, it is possible for duplicate entries to be yielded.
```

Reply:
Thank you for these edits :)